Refactor codes for alumni page

yechs opened the pull request:
Major Changes:
- Add TypeScript type checking for majority of codes
- **Need Help**: make the types more granular
- Separate alumniData from code logic

Minor improvements:
- Implement a useBaseUrl hook on the URL at the component level, so that the site can now be deployed to another `baseUrl` without changing the (previously absolute) url of images
- Fixed a bug for image alt text to overflow
- Add editorconfig for cross-platform collaboration

yechs committed:
Refactor code for alumni page

- Add TypeScript type checking for majority of codes
- **Need Help**: make the types more granular
- Separate alumniData from code logic

yechs committed:
Optimize static file url handling

Implement a useBaseUrl hook on the URL at the component level,
so that the site can now be deployed to another `baseUrl` without
changing the (previously absolute) url of images

yechs committed:
Minor improvements

- Fixed a bug for image alt text to overflow
- Add editorconfig for cross-platform collaboration

yechs commented on src/components/Members.tsx:
```diff
@@ -3,17 +3,21 @@ import styles from './Members.module.css';
+  readonly linkedin?: string;
+  readonly email?: string; // should begin with "mailto:"
+  readonly favoredLink?: "website" | "github" | "linkedin" | "email";
 };
```
I'm thinking about making these types more granular 

For example:
- `email` must begin with `mailto:` followed by the email address
- `bio` should be a string no longer than 120 characters
- `linkedin` and `github` should begin with their domain

yechs replied:
I'm still going through the TypeScript Manuals for ideas
Let's hope I do find something helpful

SamChou19815 replied:
> I'm thinking about making these types more granular 
> 
> 
> 
> For example:
> 
> - `email` must begin with `mailto:` followed by the email address
> 
> - `bio` should be a string no longer than 120 characters
> 
> - `linkedin` and `github` should begin with their domain

This is impossible to achieve in typescript

yechs replied:
> > I'm thinking about making these types more granular 
> > 
> > 
> > 
> > For example:
> > 
> > - `email` must begin with `mailto:` followed by the email address
> > 
> > - `bio` should be a string no longer than 120 characters
> > 
> > - `linkedin` and `github` should begin with their domain
> 
> This is impossible to achieve in typescript 

That's sad : (
I was just reading through this issue https://github.com/microsoft/TypeScript/issues/6579 and thought if there exists a workaround for now

yechs replied:
I'll see if i can find a solution by tomorrow. 

Otherwise we'll have to leave it like this. Implementing a format check during run-time would be unnecessary (plus unwise).

SamChou19815 replied:
> I'll see if i can find a solution by tomorrow. 
> 
> Otherwise we'll have to leave it like this. Implementing a format check during run-time would be unnecessary (plus unwise).

In general that will require a dependent type system, and that's usually not what most of the programming language can provide.

yechs replied:
> 
> In general that will require a dependent type system, and that's usually not what most of the programming language can provide.

Thanks for pointing that out! I'm still pretty new to these PL concepts and will definitely look more into things related with type systems.

SamChou19815 commented on src/components/AlumniTimeline.tsx:
```diff
@@ -1,18 +1,24 @@
       <ul className={styles.timeline}>
         {
-          alumni.map(({ classOf, members }) => (
+          alumni.map(({ classOf, members }: alumniOfYear): ReactElement => (
```
These type annotations are not really necessary. TypeScript can use type inference to figure out from the context

yechs replied:
Thanks for pointing that out, will remove them in the next commit

SamChou19815 commented:
You can also follow the instructions in https://github.com/facebook/docusaurus/pull/2687 to setup type declaration files to avoid red squiggly lines below `@docusaurus/*`, etc

yechs commented:
> You can also follow the instructions in https://github.com/facebook/docusaurus/pull/2687 to setup type declaration files to avoid red squiggly lines below `@docusaurus/*`, etc

Wow thanks I was just Googling for that yesterday!

yechs committed:
Improve types

- Remove unneeded type annotations (which can be figured out with type inference, thanks for @SamChou19815 for pointing out)
- Add a few comments to make the Member type clearer for new maintainers
- Remove `null` entries from alumniData

yechs committed:
Adopt @docusaurus/module-type-aliases

As stated in https://github.com/facebook/docusaurus/pull/2687

yechs committed:
WIP: Support ESLint

Using AirBnb style with support to React JSX and TypeScript

Yet ESLint still gives error "import/no-unresolved: Unable to resolve path to module '@theme/Layout'."

Also, there are lots of other errors (regarding codestyle) that I will fix later with another PR
